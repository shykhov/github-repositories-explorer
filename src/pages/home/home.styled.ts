import styled from 'styled-components';

export const Root = styled.div`
  margin: 50px;
  display: flex;
`;

export const ControlsWrapper = styled.div`
  width: 400px;
  margin-left: 50px;

  & > div:first-of-type {
    margin-bottom: 25px;
  }
`;
