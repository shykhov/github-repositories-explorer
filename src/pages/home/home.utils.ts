import isPlainObject from 'lodash/isPlainObject';
import isEmpty from 'lodash/isEmpty';
import isNil from 'lodash/isNil';

interface SearchParams {
  [repositorySearchValue: string]: string | null;
  userLogin: string | null;
  page: string | null;
  rowsPerPage: string | null;
}

export const prepareSearchParams = (params: SearchParams): string => {
  if (isPlainObject(params) && !isEmpty(params)) {
    const result = Object.keys(params).reduce((accumulator, currentValue, index) => {
      const newValue = currentValue || '';
      const paramsValue: string | null = params[newValue];
      if (isNil(currentValue) || isNil(paramsValue) || isEmpty(paramsValue)) {
        return accumulator;
      }

      return `${accumulator}${accumulator.length === 1 ? '' : '&'}${currentValue}=${paramsValue || ''}`;
    }, '?');

    return result.length === 1 ? '' : result;
  }

  return '';
};

interface QueryParams {
  [name: string]: string | null;
  owner: string | null;
}

export const prepareQueryParams = (params: QueryParams): string => {
  if (isPlainObject(params) && !isEmpty(params)) {
    const result = Object.keys(params).reduce((accumulator, currentValue) => {
      const newValue = currentValue || '';
      const paramsValue: string | null = params[newValue];
      if (isNil(currentValue) || isNil(paramsValue) || isEmpty(paramsValue)) {
        return accumulator;
      }

      return `${accumulator}${currentValue}:${paramsValue}`;
    }, '');

    return result;
  }

  return '';
};
