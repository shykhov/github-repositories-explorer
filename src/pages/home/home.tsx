import React, { useCallback, useMemo, ChangeEvent, FC } from 'react';
import { useLazyQuery } from '@apollo/react-hooks';
import { ApolloError } from 'apollo-client';
import debounce from 'lodash/debounce';
import { History } from 'history';
import { RouteComponentProps } from 'react-router-dom';

import { Root, ControlsWrapper } from './home.styled';
import { useUrlQuery } from '../../hooks';
import { FETCH_REPOSTORIES, FETCH_USERS } from '../../api/queries';
import {
  REPOSITORIES_PATHNAME,
  REPOSITORIES_PER_PAGE,
  DEFAULT_DEBOUNCE_TIMEOUT,
  USER_LOGIN_PARAMETER,
  REPOSITORY_NAME_QUERY_PARAMETER,
  REPOSITORY_NAME_QUERY_KEY,
  DEFAULT_REPOSITORY_SORT_QUERY,
  PAGINATION_DIRECTION,
} from '../../constants';
import {
  generateCursor,
  formatRepositories,
  formatUsers,
  findCurrentUser,
  prepareQueryString,
  prepareSearchParams,
  RepositoryOptions,
  UserOptions,
  SelectOption,
} from '../../utils';
import { RepositoriesSearchInput } from '../../components/repositories-search-input';
import { RepositoryTable } from '../../components/repository-table';
import { UserSearchSelect } from '../../components/user-search-select';

export type ApolloLazyQueryResult<T> = {
  data: T;
  called: boolean;
  error?: ApolloError | undefined;
  loading: boolean;
};

export interface HomeProps {
  getRepositories(variables: Record<string, unknown>): void;
  repositoriesQuery: ApolloLazyQueryResult<RepositoryOptions>;
  usersQuery: ApolloLazyQueryResult<UserOptions>;
  getUsers(variables: Record<string, unknown>): void;
  history: History;
}

export const Home: FC<HomeProps> = props => {
  const { getRepositories, repositoriesQuery, usersQuery, getUsers, history } = props;
  const query = useUrlQuery();
  const [userLoginSearchValue, setUserLoginSearchValue] = React.useState('');
  const [repositoryNameSearchValue, setRepositoryNameSearchValue] = React.useState('');
  const [page, setPage] = React.useState(0);

  const userLoginParameter = query.get(USER_LOGIN_PARAMETER);
  const repositoryNameSearchParameter = query.get(REPOSITORY_NAME_QUERY_PARAMETER);

  const userOptions = useMemo(() => formatUsers(usersQuery.data), [usersQuery.data]);

  const currentUser = useMemo(() => findCurrentUser(userOptions, userLoginParameter), [
    userOptions,
    userLoginParameter,
  ]);

  const repositoriesData = useMemo(() => formatRepositories(repositoriesQuery.data), [repositoriesQuery.data]);

  const handleChangePage = (event: ChangeEvent, newPage: number) => {
    const {
      repositoryCount,
      pageInfo: { endCursor, startCursor },
    } = repositoriesData;

    const { first, last } = PAGINATION_DIRECTION;

    if (newPage * REPOSITORIES_PER_PAGE + REPOSITORIES_PER_PAGE <= repositoryCount) {
      getRepositories({
        variables: {
          queryString: prepareQueryString({
            [REPOSITORY_NAME_QUERY_KEY]: repositoryNameSearchParameter,
            user: userLoginParameter,
            ...DEFAULT_REPOSITORY_SORT_QUERY,
          }),
          [newPage > page || newPage === 0 ? first : last]: REPOSITORIES_PER_PAGE,
          ...generateCursor({ newPage, oldPage: page, endCursor, startCursor }),
        },
      });
    }

    setPage(newPage);
  };

  const getNewUsers = (userLogin: string) => {
    getUsers({
      variables: {
        name: userLogin,
      },
    });
  };

  const debouncedHandleUserSelectInputChange = useCallback(debounce(getNewUsers, DEFAULT_DEBOUNCE_TIMEOUT), []);

  const handleUserSelectInputChange = (userLogin: string) => {
    if (userLogin) {
      setUserLoginSearchValue(userLogin);
      debouncedHandleUserSelectInputChange(userLogin);
    }
  };

  const handleSubmitInputSearch = (repositoryName: string): void => {
    setPage(0);

    const currentRepositoryName: string = repositoryName || repositoryNameSearchValue;

    history.push({
      pathname: REPOSITORIES_PATHNAME,
      search: prepareSearchParams({
        [REPOSITORY_NAME_QUERY_PARAMETER]: currentRepositoryName,
        [USER_LOGIN_PARAMETER]: userLoginParameter,
      }),
    });
  };

  const debouncedSubmitInputSearch = useCallback(debounce(handleSubmitInputSearch, DEFAULT_DEBOUNCE_TIMEOUT), [
    userLoginParameter,
    repositoryNameSearchParameter,
  ]);

  const handleSearchRepositoryInputChange = (event: ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;

    if (value) {
      debouncedSubmitInputSearch(value.trim());
    } else {
      debouncedSubmitInputSearch.cancel();

      history.push({
        pathname: REPOSITORIES_PATHNAME,
        search: prepareSearchParams({
          [USER_LOGIN_PARAMETER]: userLoginParameter,
        }),
      });
    }

    setRepositoryNameSearchValue(value);
  };

  const handleSelectChange = (user: SelectOption) => {
    setPage(0);
    history.push({
      pathname: REPOSITORIES_PATHNAME,
      search: prepareSearchParams({
        [REPOSITORY_NAME_QUERY_PARAMETER]: repositoryNameSearchParameter,
        [USER_LOGIN_PARAMETER]: user ? user.value : '',
      }),
    });
  };

  React.useEffect(() => {
    if (repositoryNameSearchParameter || userLoginParameter) {
      getRepositories({
        variables: {
          first: REPOSITORIES_PER_PAGE,
          queryString: prepareQueryString({
            [REPOSITORY_NAME_QUERY_KEY]: repositoryNameSearchParameter,
            user: userLoginParameter,
            ...DEFAULT_REPOSITORY_SORT_QUERY,
          }),
        },
      });
    }
  }, [getRepositories, userLoginParameter, repositoryNameSearchParameter]);

  return (
    <Root>
      <RepositoryTable
        error={Boolean(repositoriesQuery.error)}
        loading={repositoriesQuery.loading}
        called={repositoriesQuery.called}
        repositoriesData={repositoriesData}
        page={page}
        handleChangePage={handleChangePage}
      />
      <ControlsWrapper>
        <RepositoriesSearchInput
          value={repositoryNameSearchValue || repositoryNameSearchParameter}
          loading={repositoriesQuery.loading}
          handleSearchInputChange={handleSearchRepositoryInputChange}
          debouncedSubmitInputSearch={debouncedSubmitInputSearch}
        />
        <UserSearchSelect
          options={userOptions}
          inputValue={userLoginSearchValue}
          loading={usersQuery.loading}
          onSelectChange={handleSelectChange}
          onInputChange={handleUserSelectInputChange}
          value={currentUser}
        />
      </ControlsWrapper>
    </Root>
  );
};

export const HomeContainer: FC<RouteComponentProps> = ({ history }) => {
  const [getRepositories, repositoriesQuery] = useLazyQuery(FETCH_REPOSTORIES);
  const [getUsers, usersQuery] = useLazyQuery(FETCH_USERS);

  return (
    <Home
      history={history}
      usersQuery={usersQuery}
      getUsers={getUsers}
      repositoriesQuery={repositoriesQuery}
      getRepositories={getRepositories}
    />
  );
};
