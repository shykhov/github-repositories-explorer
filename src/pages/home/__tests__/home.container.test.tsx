import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import { RouteComponentProps } from 'react-router-dom';
import { MockedProvider } from '@apollo/react-testing';
import 'jest-styled-components';

import { HomeContainer } from '../home';

jest.mock('../../../hooks', () => ({
  useUrlQuery: jest.fn(),
}));

describe('<HomeContainer />', () => {
  let shallowRenderedComponent: ShallowWrapper;
  let propsMock: RouteComponentProps;

  beforeEach(() => {
    shallowRenderedComponent = shallow(
      <MockedProvider addTypename={false}>
        <HomeContainer {...propsMock} />
      </MockedProvider>,
    )
      .dive()
      .dive()
      .dive()
      .dive();
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should render HomeContainer', () => {
    expect(shallowRenderedComponent).toMatchSnapshot();
  });
});
