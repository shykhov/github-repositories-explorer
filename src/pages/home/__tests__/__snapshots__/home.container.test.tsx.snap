// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HomeContainer /> should render HomeContainer 1`] = `
<Home
  getRepositories={[Function]}
  getUsers={[Function]}
  repositoriesQuery={
    Object {
      "called": false,
      "data": undefined,
      "loading": false,
      "networkStatus": 7,
    }
  }
  usersQuery={
    Object {
      "called": false,
      "data": undefined,
      "loading": false,
      "networkStatus": 7,
    }
  }
/>
`;
